### PR TITLE
[Refactor][Enhancement] Refactor DataStreamRecvr::SenderQueue

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [x] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The way `SenderQueue` is used in the two execute engines is very different. In the non-pipeline engine, It should block and wait if there is no new chunk arriving when getting chunk, `std::mutex` and `std::condition_variable` are introduced for this. But in the pipeline engine, the execution threads are shared and it is necessary to try to avoid blocking caused by lock competition.

The current implementation is not customized for the two scenarios and all the logic is mixed together, which is very unoptimized and has additional overhead.

In this PR, I try to refactor the implementation of `SenderQueue`, the main changes are as follows:

1. Add `SenderQueue` as a base class, based on which `PipelineSenderQueue` and `NonPipelineSenderQueue` are implemented respectively;
2. The original `SenderQueue` implementation is moved to `NonPipelineSenderQueue`, and member variables and related interfaces that will not be used in the non-pipeline engine are moved.
3. The `PipelineSenderQueue` is specially implemented for the pipeline engine. In order to avoid blocking the pipeline execution thread, it will use lock-free structures as much as possible. For this reason, a lock-free queue is introduced from https://github.com/cameron314/concurrentqueue. From the public benchmark information(https://moodycamel.com/blog/2014/a-fast-general-purpose-lock-free-queue-for-c++.htm#benchmarks and https://github.com/Qihoo360/evpp/blob/master/docs/benchmark_lockfree_vs_mutex_cn.md), it has good performance.

Optimizations in PipelineSenderQueue

1. if _is_pipeline_level_shuffle=true, a separate lock-free queue will be created for each driver sequence to avoid competition;

2. introduce the unplug mechanism to reduce the scheduling overhead of exchange source pipeline driver

3. introduce a lazy deserialization mechanism, move the deserialization operation from the add_chunk stage to the get_chunk stage which can reduce the latency of transmit_chunk rpc and can also reduce some uncessary computing overhead when short_circuit happens.

### CODE REVIEW

File: be/src/runtime/sender_queue.cpp [hunk 1]
```diff
@@ -0,0 +1,897 @@
+#include "runtime/sender_queue.h"
```
[the remark this answers is not in the document]
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:3]: > > these codes are write by starrocks, should add license header.
> 
> OK…btw, this PR is working in progress, I need add some comments and descriptions, you can review it later~ @imay

fix some problems and add more information, it's ready to review, PTAL @imay

File: be/src/runtime/sender_queue.cpp [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+    return Status::OK();
+}
+
+Status DataStreamRecvr::SenderQueue::_deserialize_chunk(const ChunkPB& pchunk, vectorized::Chunk* chunk,
+                                                        faststring* uncompressed_buffer) {
+    if (pchunk.compress_type() == CompressionTypePB::NO_COMPRESSION) {
+        SCOPED_TIMER(_recvr->_deserialize_chunk_timer);
+        TRY_CATCH_BAD_ALLOC({
+            serde::ProtobufChunkDeserializer des(_chunk_meta);
+            StatusOr<vectorized::Chunk> res = des.deserialize(pchunk.data());
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:74]: ASSIGN_OR_RETURN ?

File: be/src/runtime/sender_queue.cpp [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+
+        closure->Run();
+    }
+
+    return Status::OK();
+}
+
+bool DataStreamRecvr::NonPipelineSenderQueue::has_chunk() {
+    std::lock_guard<Mutex> l(_lock);
+    if (_is_cancelled) {
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:148]: Simplify it one return statement?

File: be/src/runtime/sender_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+    _data_arrival_cv.notify_all();
+
+    {
+        std::lock_guard<Mutex> l(_lock);
+        clean_buffer_queues();
+    }
+}
+
+void DataStreamRecvr::NonPipelineSenderQueue::close() {
+    {
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:468]: Is the outer bracket useless?

File: be/src/runtime/sender_queue.h [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            _size.store(rhs._size);
+            return *this;
+        }
+
+        inline bool empty() const { return _size.load() == 0; }
+
+        inline bool size() const { return _size.load(); }
+
+        void enqueue(ChunkItem&& item) {
+            _chunks.enqueue(std::move(item));
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.h:192]: Why not use `ConcurrentQueue::size()`, cause `_size` may be unconsistent with `_chunks.size()`?

File: be/src/util/spinlock.h [hunk 6]
```diff
@@ -65,7 +65,7 @@ class SpinLock {
     // To try to get the best of everything, we will busy spin for a while before
     // yielding to another thread.
     // TODO: how do we set this.
-    static const int NUM_SPIN_CYCLES = 70;
+    static const int NUM_SPIN_CYCLES = 4000;
```
Comment (mofeiatwork) [inline be/src/util/spinlock.h:68]: Why more spins?

**Comment by Pslydhh** [comment]
As I know, concurrentqueue is not linearizable and not sequentially consistent, is this affect its usage in starrocks?

File: be/src/runtime/sender_queue.cpp [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+        }
+
+        if (_buffered_chunk_queues.find(be_number) == _buffered_chunk_queues.end()) {
+            _buffered_chunk_queues[be_number] = phmap::flat_hash_map<int64_t, ChunkQueue>();
+        }
+    }
+
+    ScopedTimer<MonotonicStopWatch> wait_timer(_recvr->_sender_wait_lock_timer);
+    {
+        std::lock_guard<Mutex> l(_lock);
```
Comment (ZiheLiu) [inline be/src/runtime/sender_queue.cpp:318]: Here unlock and lock immediately, maybe we shouldn't unlock in L314?

File: be/src/runtime/sender_queue.cpp [hunk 8]
```diff
@@ -0,0 +1,899 @@
...
+        return Status::Cancelled("Cancelled SenderQueueForPipeline::get_chunk");
+    }
+
+    auto& chunk_queue = _is_pipeline_level_shuffle ? _chunk_queues[driver_sequence] : _chunk_queues[0];
+
+    if (chunk_queue.empty()) {
+        return Status::OK();
+    }
+    ChunkItem item;
+    if (!chunk_queue.try_dequeue(item)) {
```
Comment (ZiheLiu) [inline be/src/runtime/sender_queue.cpp:517]: There may be race condition between `get_chunk()` and `cancel()`. For example, 
1. `get_chunk()` finds `_is_cancelled` is false and is continue.
2. `cancel()` clears `_chunk_queues`.
3. `get_chunk()` gets `_chunk_queues[driver_sequence]` and exceeds index boundary.

Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:517]: > There may be race condition between `get_chunk()` and `cancel()`. For example,
> 
> 1. `get_chunk()` finds `_is_cancelled` is false and is continue.
> 2. `cancel()` clears `_chunk_queues`.
> 3. `get_chunk()` accesses `_chunk_queues[driver_sequence]` and exceeds index boundary.

yes, you're right, I think we can avoid it by clearing _chunk_queues in desctructor instead of cancel

File: be/src/util/spinlock.h (hunk 6 above)
Comment (silverbullet233) [reply be/src/util/spinlock.h:68]: > Why more spins?

The previous value is too conservative, and it is easy to cause unnecessary yields. I see that this value in many spinlock implementations are much larger than it, such as 4000 in folly:: MicroSpinLock, 1000 in absl: SpinLock and kudu

**Comment by silverbullet233** [comment]
> As I know, concurrentqueue is not linearizable and not sequentially consistent, is this affect its usage in starrocks?

I don't think this will have any effect, in our scenario as long as it is guaranteed to be eventually consistent it should be fine

File: be/src/runtime/sender_queue.h [hunk 9]
```diff
@@ -0,0 +1,230 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+#include "column/vectorized_fwd.h"
+#include "runtime/data_stream_recvr.h"
+#include "serde/protobuf_serde.h"
+#include "util/moodycamel/concurrentqueue.h"
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.h:7]: Should the third-party file be separated included?

File: be/src/runtime/sender_queue.h [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    virtual Status get_chunk(vectorized::Chunk** chunk, const int32_t driver_sequence = -1) = 0;
+
+    // check if data has come, work with try_get_chunk.
+    virtual bool has_chunk() = 0;
+
+    // Probe for chunks, because _chunk_queue maybe empty when data hasn't come yet.
+    virtual bool try_get_chunk(vectorized::Chunk** chunk) = 0;
+
+    // add chunks to this sender queue if this stream has not been cancelled
+    virtual Status add_chunks(const PTransmitChunkParams& request, ::google::protobuf::Closure** done) = 0;
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.h:32]: `put` and `get` may be more nice name ?

File: be/src/runtime/sender_queue.h [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    Status _build_chunk_meta(const ChunkPB& pb_chunk);
+
+    Status _deserialize_chunk(const ChunkPB& pchunk, vectorized::Chunk* chunk, faststring* uncompressed_buffer);
+
+    DataStreamRecvr* _recvr = nullptr;
+
+    serde::ProtobufChunkMeta _chunk_meta;
+};
+
+class DataStreamRecvr::NonPipelineSenderQueue : public DataStreamRecvr::SenderQueue {
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.h:56]: Could the two derived class be moved to `.cpp` file? Nobody need to know the implementation but only the public interface?

File: be/src/runtime/sender_queue.h [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+    std::atomic<bool> _is_cancelled{false};
+    std::atomic<int> _num_remaining_senders;
+
+    typedef SpinLock Mutex;
+    Mutex _lock;
+
+    // if _is_pipeline_level_shuffle=true, we will create a queue for each driver sequence to avoid competition
+    // otherwise, we will only use the first item
+    std::vector<ChunkQueue> _chunk_queues;
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.h:203]: How about embed these three `vector` into a struct ?

File: be/src/runtime/sender_queue.cpp [hunk 13]
```diff
@@ -0,0 +1,957 @@
...
+
+    if (_is_cancelled) {
+        return Status::Cancelled("Cancelled SenderQueue::get_chunk");
+    }
+
+    if (_chunk_queue.empty()) {
+        return Status::OK();
+    }
+
+    *chunk = _chunk_queue.front().chunk_ptr.release();
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:123]: Is it safe to release at here?

File: be/src/runtime/sender_queue.cpp [hunk 14]
```diff
@@ -0,0 +1,957 @@
...
+            ChunkItem item{chunk_bytes, std::move(chunk), nullptr};
+            chunks.emplace_back(std::move(item));
+            total_chunk_bytes += chunk_bytes;
+        }
+        COUNTER_UPDATE(_recvr->_bytes_received_counter, total_chunk_bytes);
+    }
+
+    wait_timer.start();
+    {
+        std::lock_guard<Mutex> l(_lock);
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:269]: Shall we write a util like `TimedLockGuard` ?

Comment (mofeiatwork) [reply be/src/runtime/sender_queue.cpp:269]: And print a warning log if wait a lock for too long ?

File: be/src/runtime/sender_queue.cpp [hunk 15]
```diff
@@ -0,0 +1,957 @@
...
+}
+
+Status DataStreamRecvr::NonPipelineSenderQueue::add_chunks_and_keep_order(const PTransmitChunkParams& request,
+                                                                          ::google::protobuf::Closure** done) {
+    bool use_pass_through = request.use_pass_through();
+    DCHECK(request.chunks_size() > 0 || use_pass_through);
+    const int32_t be_number = request.be_number();
+    const int32_t sequence = request.sequence();
+
+    ScopedTimer<MonotonicStopWatch> wait_timer(_recvr->_sender_wait_lock_timer);
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:301]: How about put it into the following bracket?

File: be/src/runtime/sender_queue.cpp [hunk 16]
```diff
@@ -0,0 +1,957 @@
...
+                                                        ::google::protobuf::Closure** done) {
+    bool use_pass_through = request.use_pass_through();
+    DCHECK(request.chunks_size() > 0 || use_pass_through);
+    if (_is_cancelled || _num_remaining_senders <= 0) {
+        return Status::OK();
+    }
+
+    ScopedTimer<MonotonicStopWatch> wait_timer(_recvr->_sender_wait_lock_timer);
+    {
+        std::lock_guard<Mutex> l(_lock);
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:580]: Is it efficient to use a spinlock here?

File: be/src/runtime/sender_queue.cpp [hunk 17]
```diff
@@ -0,0 +1,957 @@
...
+    const int32_t be_number = request.be_number();
+    const int32_t sequence = request.sequence();
+    if (_is_cancelled) {
+        return Status::OK();
+    }
+
+    {
+        std::lock_guard<Mutex> l(_lock);
+
+        if (_max_processed_sequences.find(be_number) == _max_processed_sequences.end()) {
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:693]: How about use the `phmap::lazy_emplace` interface?

File: be/src/runtime/sender_queue.cpp [hunk 18]
```diff
@@ -0,0 +1,957 @@
...
+        COUNTER_UPDATE(_recvr->_bytes_received_counter, total_chunk_bytes);
+    }
+
+    if (_is_cancelled) {
+        return Status::OK();
+    }
+
+    wait_timer.start();
+    {
+        std::lock_guard<Mutex> l(_lock);
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:777]: Would this lock section be too large?

File: be/src/runtime/sender_queue.cpp [hunk 19]
```diff
@@ -0,0 +1,957 @@
...
+            ++max_processed_sequence;
+        }
+    }
+    return Status::OK();
+}
+
+void DataStreamRecvr::PipelineSenderQueue::decrement_senders(int be_number) {
+    {
+        std::lock_guard<Mutex> l(_lock);
+        if (_sender_eos_set.find(be_number) != _sender_eos_set.end()) {
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:845]: Why not use the `parallel_flat_set`?

File: be/src/runtime/sender_queue.cpp [hunk 20]
```diff
@@ -0,0 +1,957 @@
...
+    size_t chunk_num = _chunk_queues[index].size_approx();
+    // introduce an unplug mechanism similar to scan operator to reduce scheduling overhead
+
+    // 1. in the unplug state, return true if there is a chunk, otherwise return false and exit the unplug state
+    if (_unpluging[index]) {
+        if (chunk_num > 0) {
+            return true;
+        }
+        _unpluging[index] = false;
+        return false;
```
Comment (mofeiatwork) [inline be/src/runtime/sender_queue.cpp:930]: Why return here?

File: be/src/runtime/sender_queue.cpp (hunk 14 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:269]: > Shall we write a util like `TimedLockGuard` ?

I think it is necessary

File: be/src/runtime/sender_queue.cpp (hunk 20 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:930]: > Why return here?

there is no chunk any more

File: be/src/runtime/sender_queue.cpp (hunk 13 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:123]: > Is it safe to release at here?

These are all existing codes in original SenderQueue, and they should be fine.

File: be/src/runtime/sender_queue.cpp (hunk 18 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:777]: > Would this lock section be too large?

I think this may be a potential problem, but we have no way to remove this lock at present…after this pr, I tend to try streaming rpc and let it solve the problem of maintaining order…

File: be/src/runtime/sender_queue.h (hunk 11 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.h:56]: > Could the two derived class be moved to `.cpp` file? Nobody need to know the implementation but only the public interface?

`PipelineSenderQueue` has serveral interfaces that are not common, such as has_output and short_circuit, so the caller still needs to be aware of the definition…

**Comment by silverbullet233** [comment]
for the convenience of review, I split this PR into four:
1. https://github.com/StarRocks/starrocks/pull/9711
2. https://github.com/StarRocks/starrocks/pull/9719
3. https://github.com/StarRocks/starrocks/pull/9740
4. https://github.com/StarRocks/starrocks/pull/9741

File: be/src/runtime/sender_queue.cpp (hunk 16 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:580]: > Is it efficient to use a spinlock here?

I think it's ok, because we won't hold the lock for a long time

File: be/src/common/config.h [hunk 21]
```diff
@@ -660,7 +660,7 @@ CONF_Int64(pipeline_prepare_thread_pool_queue_size, "102400");
 // The buffer size of SinkBuffer.
 CONF_Int64(pipeline_sink_buffer_size, "64");
 // The degree of parallelism of brpc.
-CONF_Int64(pipeline_sink_brpc_dop, "8");
+CONF_Int64(pipeline_sink_brpc_dop, "64");
```
Comment (fzhedu) [inline be/src/common/config.h:663]: is this change related to doc? for example the default value in a config file.

Comment (silverbullet233) [reply be/src/common/config.h:663]: > is this change related to doc? for example the default value in a config file.

I did't find it in our user docs and default config file, this seems to be an advaned configuration
